### PR TITLE
fixes #3859 clip_norm works with Embedding layer

### LINKED ISSUE #3859: clipnorm doesn't work with Embedding
I'm getting a Traceback every time "clipnorm" is used in NN with Embedding layer.
Here is a simple script where the problem is obvious:

``` python
import numpy as np
from keras.layers import Input, Embedding
from keras.optimizers import Adam
from keras.models import Model

input_layer = Input(shape = (1,) )

embedding = Embedding(input_dim = 1,
                      output_dim = 1)(input_layer)

model = Model(input = input_layer, output = embedding)

model.compile(optimizer = Adam(clipnorm = 1.0), loss = 'mse')

X = np.array([[1]])
Y = np.array([[[0.5]]])
model.fit(X, Y, nb_epoch = 1)
```

Failure:

``` shell
I tensorflow/core/common_runtime/gpu/gpu_device.cc:867] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:01:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:867] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX TITAN X, pci bus id: 0000:02:00.0)
Traceback (most recent call last):
  File "./clipnorm-bug.py", line 20, in <module>
    model.fit(X, Y, nb_epoch = 1)
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/training.py", line 1079, in fit
    self._make_train_function()
  File "/usr/local/lib/python3.5/dist-packages/keras/engine/training.py", line 696, in _make_train_function
    self.total_loss)
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 379, in get_updates
    grads = self.get_gradients(loss, params)
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 71, in get_gradients
    grads = [clip_norm(g, self.clipnorm, norm) for g in grads]
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 71, in <listcomp>
    grads = [clip_norm(g, self.clipnorm, norm) for g in grads]
  File "/usr/local/lib/python3.5/dist-packages/keras/optimizers.py", line 9, in clip_norm
    g = K.switch(n >= c, g * c / n, g)
TypeError: unsupported operand type(s) for *: 'IndexedSlices' and 'float'
```

Keras version is 1.1.0, TensorFlow is 0.10rc

clipvalue on the other hand works fine.

### PR DESCRIPTION
**Pull request by kracwarlock**
- made clip_norm implementation backend-specific
 - it now works when gradients are TensorFlow IndexedSlices
 - added test to confirm that it works
 - code to reproduce the bug (small change from #3859's code because that had other bugs as well):
```
import numpy as np
from keras.layers import Input, Embedding
from keras.optimizers import Adam
from keras.models import Model

input_layer = Input(shape = (1,) )

embedding = Embedding(input_dim = 1,
                              output_dim = 1)(input_layer)

model = Model(input = input_layer, output = embedding)

model.compile(optimizer = Adam(clipnorm = 1.0), loss = 'mse')

X = np.array([[0]])
Y = np.array([[[0.5]]])
model.fit(X, Y, nb_epoch = 1)

```

### CODE REVIEW

**Comment by fchollet** [comment]
For the time being we will not be adding new APIs to the backend. So a better fix here would be to refactor `clip_norm` to have a slightly different behavior based on the backend. In general, having backend-specific code is something we strongly seek to avoid, but in this case it would be a temporary fix.

**Comment by kracwarlock** [comment]
@fchollet i have refactored `clip_norm`
looks messy but if we are not to modify anything in the backend then is this acceptable?

**Comment by fchollet** [comment]
Please add a test to check that it works (packaged with the `Embedding` test, run only when the backend is TF).

**Comment by kracwarlock** [comment]
- added test that runs only with TF backend
 - confirmed that this test fails without this PR's fix
Done though I wasn't sure if this is the desired way of writing backend-specific tests! Let me know if you want changes.

File: tests/keras/layers/test_embeddings.py [hunk 1]
```diff
@@ -13,5 +16,14 @@ def test_embedding():
                expected_output_dtype=K.floatx())
 
 
+if K.backend() == 'tensorflow':
+    @keras_test
+    def test_emedding_with_clipnorm():
+        model = Sequential()
+        model.add(Embedding(input_dim = 1, output_dim = 1))
```
Comment (fchollet) [inline tests/keras/layers/test_embeddings.py]: PEP8: no space around `=` in function calls.

File: tests/keras/layers/test_embeddings.py [hunk 2]
```diff
@@ -13,5 +16,14 @@ def test_embedding():
                expected_output_dtype=K.floatx())
 
 
+if K.backend() == 'tensorflow':
+    @keras_test
+    def test_emedding_with_clipnorm():
```
Comment (fchollet) [inline tests/keras/layers/test_embeddings.py]: "embedding"

File: tests/keras/layers/test_embeddings.py [hunk 3]
```diff
@@ -13,5 +16,14 @@ def test_embedding():
                expected_output_dtype=K.floatx())
 
 
+if K.backend() == 'tensorflow':
```
Comment (fchollet) [inline tests/keras/layers/test_embeddings.py]: Use `@pytest.mark.skipif(K.backend() != 'tensorflow', reason='Requires TF backend')`

Comment (kracwarlock) [inline tests/keras/layers/test_embeddings.py]: fixed in 4584734

File: tests/keras/layers/test_embeddings.py (hunk 2 above)
Comment (kracwarlock) [inline tests/keras/layers/test_embeddings.py]: fixed in 4584734

File: tests/keras/layers/test_embeddings.py (hunk 1 above)
Comment (kracwarlock) [inline tests/keras/layers/test_embeddings.py]: fixed in 4584734

**Comment by kracwarlock** [comment]
Hey @fchollet I made the changes a few days ago. Any updates?

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.

**Comment by maremoto** [comment]
Hi,
Any update of this patch being available in the main Keras branch and the pip release?
I am using Keras 2.0 now and clip_norm is not working yet with TensorFlow (unsupported operand type(s) for *: 'IndexedSlices' and 'float').
Thanks a lot.

**Comment by kracwarlock** [comment]
Before I make a PR for the new keras version can @fchollet clarify if he
will accept
- changes to the backend as was in the initial commits in this PR
or
- the backend specific code solution he suggested later and which he
reviewed

**Comment by n1kt0** [comment]
this code works with the new version of keras
modified file is: "/path/to/keras/optimizers.py"

```
def clip_norm(g, c, n):
    if c > 0:
        if K.backend() == 'tensorflow':
            import tensorflow as tf
            import copy
            condition = n >= c
            then_expression = tf.scalar_mul(c / n, g)
            else_expression = g

            if hasattr(then_expression, 'get_shape'):
                g_shape = copy.copy(then_expression.get_shape())
            elif hasattr(then_expression, 'dense_shape'):
                g_shape = copy.copy(then_expression.dense_shape)
            if condition.dtype != tf.bool:
                condition = tf.cast(condition, 'bool')
            g = K.tensorflow_backend.control_flow_ops.cond(condition,
                                           lambda: then_expression,
                                           lambda: else_expression)
            if hasattr(then_expression, 'get_shape'):
                g.set_shape(g_shape)
            elif hasattr(then_expression, 'dense_shape'):
                g._dense_shape = g_shape
        else:
            g = K.switch(n >= c, g * c / n, g)
    return g
```

**Comment by danieljl** [comment]
@kracwarlock Could you please resubmit the PR to `master` as suggested by @fchollet?

**Pull request closed without merging**